pkg/handler/handler.go

```diff
@@ -33,22 +65,93 @@ func New(b *storage.BucketHandle, cacheDir, listenAddress string) *http.Server {
 	}
 }
 
-func (pbh *prowBucketHandler) router(resp http.ResponseWriter, r *http.Request) {
-	if strings.HasPrefix(r.URL.String(), "/logs") {
-		pbh.handleLogRequest(resp, r)
+func (pbh *prowBucketHandler) router(resp http.ResponseWriter, req *http.Request) {
+	pbh.requestedURL = req.URL.Path
+
+	if strings.HasPrefix(req.URL.String(), "/logs") {
+		pbh.handleLogRequest(resp, req)
+		return
+	}
+
+	if req.URL.Path != "/" {
+		http.NotFound(resp, req)
 		return
 	}
-	resp.WriteHeader(http.StatusNotFound)
+
+	if !isAuthenticated(req) {
+		page, _ := ioutil.ReadFile("pkg/handler/login.html")
+		fmt.Fprintf(resp, string(page))
+	}
 }
 
-func (pbh *prowBucketHandler) handleLogRequest(resp http.ResponseWriter, r *http.Request) {
-	log.Printf("Got request for %s", r.URL.Path)
-	if !pathChecker.MatchString(r.URL.Path) {
+func (pbh *prowBucketHandler) handleLogRequest(resp http.ResponseWriter, req *http.Request) {
+	log.Printf("Got request for %s", req.URL.Path)
+	if !pathChecker.MatchString(req.URL.Path) {
 		resp.WriteHeader(http.StatusForbidden)
 		return
 	}
 
-	bucketPath := strings.Replace(r.URL.Path, "/logs", "pr-logs/pull", 1)
+	// check if the requested repo is private.
+	// if it is private, point to the login page.
+	match := repoName.FindStringSubmatch(req.URL.Path)
+	var repo string
+	if match != nil && len(match) == 1 {
+		repo = match[0] // will be a string of the form /logs/org_repo
+		repo = strings.Replace(repo[6:], "_", "/", -1)
+	}
+
+	if !contains(pbh.publicRepos, repo) {
+		if !isAuthenticated(req) {
+			page, _ := ioutil.ReadFile("pkg/handler/login.html")
+			fmt.Fprintf(resp, string(page))
+			return
+		}
+	}
+
+	pbh.showLogs(resp, req)
+}
+
+// issueSession issues a cookie session after successful Github login
+func (pbh *prowBucketHandler) issueSession() http.Handler {
+	fn := func(resp http.ResponseWriter, req *http.Request) {
+		ctx := req.Context()
+		githubUser, err := github.UserFromContext(ctx)
+		if err != nil {
+			http.Error(resp, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		isMember, _, err := pbh.githubClient.Organizations.IsMember(ctx, pbh.org, *githubUser.Login)
+		if err != nil {
+			http.Error(resp, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		if !isMember {
+			_, err := fmt.Fprintf(resp, `<p> Hey, %s</p>`, *githubUser.Login)
+			if err != nil {
+				http.Error(resp, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			fmt.Fprintf(resp, `<p>Only members of the %s organisation are allowed to view this page. Looks like you are not a member. :(</p>`, pbh.org)
+			return
+		}
+
+		session := sessionStore.New(sessionName)
+		session.Values[sessionUserKey] = *githubUser.ID
+		session.Save(resp)
```

Comment:
So..it _did_ work correctly. It seemed like it wasn't working because I was saving some stuff in `pbh`, which caused concurrency issues. Fixed it now.

> Right now there is nothing that is unique per session or am I understanding it wrong?

```go
func isAuthenticated(req *http.Request) bool {
	if _, err := sessionStore.Get(req, sessionName); err == nil {
		return true
	}
	return false
}
```

Turns out each session is unique because it's each session _per request_. Tested this locally as well. :+1: 

And we also require that `sessionName` be a constant value because if this isn't constant and it is idenitified by a github user (say their `username`)  there is actually no way to identify the current user's `username` without making the person log in again.

```go
ctx := req.Context()
￼githubUser, err := github.UserFromContext(ctx)
```
This ^ won't help while trying to indentify the current user because there is no github user for _the_ current request's context.

Reply:
Makes sense, glad you figured it out!